### PR TITLE
dev-guide/api-conventions: Explain preference for single canonical phrasing

### PR DESCRIPTION
**Pull request by wking**
Proximal trigger is @deads2k's concern about ClusterVersion's `spec.desiredUpdate` being a pointer structure (violating our current API conventions) where both the nil:

```yaml
spec: {}
```

and the empty-struct:

```yaml
spec:
  desiredUpdate: {}
```

have the same semantics.  If we created desiredUpdate today, the OpenShift anti-pointer preference would have forbidden the nil case, but this section gives more motivational background for folks creating new APIs.

### CODE REVIEW

**Comment by JoelSpeed** [comment]
Changes read well to me, I'll defer to @deads2k to approve

File: dev-guide/api-conventions.md [hunk 1]
```diff
@@ -109,6 +109,24 @@ This means, that any end user fetching the resource from the API can observe tha
 This has the effect of helping our users to understand how they might be able to configure their cluster, without
 having to search for features or review API schemas within the product docs.
 
+### Only one phrasing for each idea
+
+Each idea should have a single possible expression in the API structures, without having multiple ways to say the same thing.
```
Comment (markmc) [inline dev-guide/api-conventions.md:114]: The example in the commit message would help the reader understand how the principle might be applied in the context of Kubernetes API modelling

Comment (wking) [reply dev-guide/api-conventions.md:114]: I've worked that example in with b62f09e0 -> a3f21140, but if anyone has a real-world example that doesn't violate another of our conventions, that would be useful.

File: dev-guide/api-conventions.md [hunk 2]
```diff
@@ -109,6 +109,40 @@ This means, that any end user fetching the resource from the API can observe tha
...
+From [PEP 20][pep-20]:
+
+> * There should be one-- and preferably only one --obvious way to do it.
+> * Although that way may not be obvious at first unless you're Dutch.
+
+For example, [ClusterVersion's `spec.desiredUpdate`][desiredUpdate] is a pointer (because it was added before [the anti-pointer guidelines](#pointers)).
```
Comment (JoelSpeed) [inline dev-guide/api-conventions.md:120]: I don't really like the wording here, "before the anti-pointer guidelines". It doesn't feel like the right tone of voice for an API guidelines doc. We are trying for passive voice and I think this is too active voice.

I think it would also be good to explain that it allows to representations of the same thing.

I'd suggest rewording the example to something like
```suggestion
For example, the [ClusterVersion's `spec.desiredUpdate`][desiredUpdate] is a pointer field.
This does not meet our current API guidelines (see our [pointer guidelines](#pointers)) as we advise against the use of pointers in most cases, but, in this example, it also allows the same semantic to be expressed in two different ways.

Both `nil`:
...
and the empty-struct:
...
represent the same "I do not desire an update" idea.
```

Comment (deads2k) [reply dev-guide/api-conventions.md:120]: After this is reworded, the update looks good to me.

File: dev-guide/api-conventions.md (hunk 1 above)
Comment (Miciah) [reply dev-guide/api-conventions.md:114]: I agree that a real-world example would be useful.  As it is, this new guideline seemingly proscribes only choices that are already proscribed by other guidelines.  In addition, it proscribes choices that other guidelines recommend, such as discriminated unions with empty union members.

File: dev-guide/api-conventions.md (hunk 2 above)
Comment (wking) [reply dev-guide/api-conventions.md]: Rebased onto master and rolled in @JoelSpeed 's suggestions with a3f21140 -> f2d90a07.

File: dev-guide/api-conventions.md (hunk 1 above)
Comment (wking) [reply dev-guide/api-conventions.md:114]: > In addition, it proscribes choices that other guidelines recommend, such as discriminated unions with empty union members.

Good point.  I'm unclear how [the current discriminated union pointer requirement][1] squares with [the no-pointer guideline][2].  But given that that conflict already exists today, I don't see how my proposed change is making it worse.

[1]: https://github.com/openshift/enhancements/blob/890debb43169f0458246c3e19976a4c4bdda5b30/dev-guide/api-conventions.md#writing-a-union-in-go
[2]: https://github.com/openshift/enhancements/blob/890debb43169f0458246c3e19976a4c4bdda5b30/dev-guide/api-conventions.md#pointers

Comment (JoelSpeed) [reply dev-guide/api-conventions.md:114]: The discriminated union pointer is an exception that's required, we should make a note that this is an exception to the rule I guess

**Comment by JoelSpeed** [comment]
/approve
/lgtm

/hold

This looks good, holding for now in case @wking wants to get feedback from additional reviewers, I'll let you make the call on how much bike shedding to do

**Comment by JoelSpeed** [comment]
As this has gone stale, I'm just going to let it merge as is, I think it looks good and if others have additional feedback we can follow up on it
/hold cancel
